test/charms/test_main/lib/charm.py

```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env python3
+
+import os
+import base64
+import pickle
+
+from juju.charm import (
+    CharmBase,
+)
+
+import logging
+
+logger = logging.getLogger()
+
+
+class Charm(CharmBase):
+
+    def __init__(self, framework, key, metadata):
+        super().__init__(framework, key, metadata)
+
+        # Load attributes that might have been passed via an environment
+        # variable (used for altering charm behavior during testing).
+        charm_config = os.environ.get('CHARM_CONFIG')
+        if charm_config is not None:
+            self._charm_config = pickle.loads(base64.b64decode(charm_config))
+        else:
+            self._charm_config = {}
+
+        self._state_file = self._charm_config.get('STATE_FILE')
+        self._state = {}
+        self._state['event_type_names_observed'] = list()
+
+        self.framework.observe(self.on.install, self)
+        self.framework.observe(self.on.config_changed, self)
+        self.framework.observe(self.on.update_status, self)
+        self.framework.observe(self.on.leader_settings_changed, self)
+        # Test relation events with endpoints from different
+        # sections (provides, requires, peers) as well.
+        self.framework.observe(self.on.db_relation_joined, self)
+        self.framework.observe(self.on.mon_relation_changed, self)
+        self.framework.observe(self.on.ha_relation_broken, self)
+
+    def _write_state(self):
+        """Write state variables so that the parent process can read them.
+
+        Each invocation will override the previous state which is intentional.
+        """
+        if self._state_file is not None:
+            with open(self._state_file, 'rb+') as f:
+                pickle.dump(self._state, f)
+
+    def _handler_core(self, event, deferred=False):
+        event_type_name = type(event).__name__
+        logger.debug(f'Handling event {event_type_name}')
+
+        count_called_key = f'{event_type_name}_count_called'
+        count_called = self._state.get(count_called_key)
+        if count_called is None:
+            count_called = 0
+        count_called += 1
+        self._state[count_called_key] = count_called
+
+        count_deferred_key = f'{event_type_name}_count_deferred'
+        count_deferred = self._state.get(count_deferred_key)
+        if count_deferred is None:
+            count_deferred = 0
+
+        if deferred:
+            count_deferred += 1
```

Comment:
Still a lot of generic logic in this function which might be replaced with simpler, more readable, and more reliable logic (in terms of testing) if hardcoded.

I would imagine something along those lines for example:
```
def __init__(self):
    ...
    self._state["on_install"] = []

def on_install(self, event):
    self._state["on_install"].append(event.event_type)
```

Despite being much simpler, note that this test will actually allow detecting if we call on_install with the wrong event, while the current logic doesn't.

Reply:
I simplified the test according to your feedback.

I still need to have a list of observed events (per hook invocation/Charm object lifetime) to check for the order of event handling as seen by the charm.

`        self._state['observed_event_types'] = []`

Otherwise, my test code would have introduced its own ordering by reading per-handler states one by one.